src/hufo_tunnel.1
.TH "OGLC Screensaver" 1 "" "X Version 11"
.de Ds
.Sp
.nf
..
.de De
.fi
..
.SH NAME
hufo_tunnel - tunnel saver.
.SH SYNOPSIS
.B hufo_tunnel
[\--root/-r]
[\--maxfps/-x \fInumber\fP]
[\--vsync/-y \fInumber\fP]
[\--dpms/-M \fInumber\fP]
[\--texture/-t \fInumber\fP]
[\--coarseness/-c \fInumber\fP]
[\--sinusoide/-s]
[\--no-sinusoide/-S]
[\--wireframe/-w]
[\--no-wireframe/-W]
.SH DESCRIPTION
From Jeremie Allard (\fIhttp://hufo.planet-d.net/codefr_us.php\fP):
"A rewrited version of my first OpenGL Challenge entry. Something like a rebirth :) After too much month of not participing to this challenge I've done this little entry just to thanks Otry and BlackOrb for this great new website."

Ported to Linux by Tugrul Galatali.
.SH OPTIONS
.TP 8
.B \--root
Draw on the root window.
.TP 8
.B \--maxfps \fInumber\fP
Set maximum frame rate.
.TP 8
.B \--vsync \fInumber\fP
Limit redraws to specified number of vertical refreshes.  0 - 100.  Default: 1
.TP 8
.B \--dpms \fInumber\fP
Stop rendering new frames if the display is not on.  0 - 1.  Default: 1
.TP 8
.B \--texture \fInumber\fP
.Ds
1 - Swirl (--swirl)
2 - Marble (--marble)
.De

Default: 1
.TP 8
.B \--coarseness \fInumber\fP
Coarseness of tunnel interior.  0 - 3.  Default: 0
.TP 8
.B \--sinusoide
Enable sinusoide size.  Default: False
.TP 8
.B \--wireframe
Enable wireframe.  Default: False
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.SH SEE ALSO
.BR X (1),
.SH COPYRIGHT
Copyright \(co 2000 by Jeremie Allard and Tugrul Galatali.  

hufo_tunnel is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

hufo_tunnel is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.SH AUTHOR
Jeremie Allard (Hufo / N.A.A.) <hufo@planet-d.net> and Tugrul Galatali <tugrul@galatali.com>



ChangeLog
******************************* Version 0.9.1  *********************************

	Added port of Pixel City, originally by Shamus Young,
http://code.google.com/p/pixelcity/
	Replaced methods deprecated by ImageMagick.

******************************* Version 0.9.0  *********************************

	Added port of drempels, originally by Ryan M. Geiss,
http://www.geisswerks.com/about_drempels.html
	Added new saver feedback by Tugrul Galatali.
	Added support for DPMS to idle the saver when the display isn't on.
	Fixed brown paper bag bug in hyperspace.
	Added two new presets to flocks inspired by Matt Richard's Wavy Gravy on
openprocessing.org
	Tweaked hufo_smoke to avoid flat flame effect.
	Fixed unknown argument segfault in lattice.
	Switched matrixview's main character array to vertex arrays, which helps
significantly on older nVidia hardware.
	hyperspace and drempels have command line arguments to disable use of
shaders in case they behave unexpectedly.
	Minor changes to fix builds on other platforms.

******************************* Version 0.8.3 *********************************

	New screen saver Lorenz contributed by Soeren Sonnenburg <sonne@debian.org>
	Fix for SF #2673397 by changing =='s to >='s in widthChange update logic,
otherwise non-sense values for interpolation would wreak havoc. Also changed for
xyzChange.

******************************* Version 0.8.2 *********************************

	Limit frame rate to vertical refresh by default if compiled with GLEW
and GLX_SGI_swap_control is available. Otherwise, maxfps set to 30.
	--nice eliminated and made default. nanosleep/usleep are much better 
behaved these days.
	spirographx modified to use vertex arrays and GNU extension sincosf if
available, dropping CPU usage significantly.
	Minor updates to correct issues noted by newer compilers.
	--disable-debug is now the default in configure.

******************************* Version 0.8.1 *********************************

	Pulled in vroot.h patch from Ari Pollak's Debian package.
	Added missing #include <assert.h> to hyperspace.cpp
	Image loading in matrixview was hanging for whatever reason,
		rewritten with the new Magick Wand interface.
	ImageMagick now required.
	timeval to double conversion now fixed with an explicit cast of tv_sec
		to double.
	Fixed automake files to use CLEANFILES
	Dropped SSE/3dnow stuff
	Fixed hyperspace.xml typo in dist_kdessconfig_DATA

******************************* Version 0.8.0 *********************************

	Rough sync against new Windows release, rssavers 0.1
		Except for skyrocket
		Added new saver Hyperspace

******************************* Version 0.7.6 *********************************

	Fix to matrixview.c where the body of a thread was accidentally left as a
subfunction of a function that exits while the thread is still alive.

	Added an autoconf check for ExportImagePixels since it was very recently
added (5.5.7).

******************************* Version 0.7.5 *********************************

	Removed multiline strings that are no longer supported by gcc 3.3.

	Added spirographx, a port of an OSX app/screensaver SpirographX by 
Holmes Futrell <holmes@neatosoftware.com>.
	http://homepage.mac.com/solidmag/spiro/

	Added matrixview, based on Alex Zolotov's Matrix GL. Hacked in external 
image loading and ability to modify the resolution of the matrix.

******************************* Version 0.7.4 *********************************

	Fixed bug in lattice when the variable passed to acos would be slightly
greater than 1, causing cascading nan's that made the program useless.

	Explicitly clearing buffers to avoid garbage in blurring modes.

	Lots of stupid little bug fixes.

******************************* Version 0.7.3 *********************************

	Added trails and blur options to flocks. Also added presets to spruce
things up in combination with random preset selection when running 
unconfigured.

	Fixed another cpu detection bug, damn off by one errors :)

	Added support for decimal integer arguments to -window-id. 

	Added .desktop files for preliminary support of KDE. Not sure where
the default directory is, but if it isn't specified by --with-kdessconfigdir,
the configure script checks for the existence of:

/usr/share/applnk/System/ScreenSavers

******************************* Version 0.7.2 *********************************

	Fixed bug in cpu detection code.

	Fixed a bug that included audio data in skyrocket when OpenAL wasn't 
found.

******************************* Version 0.7.1 *********************************

	Fixed a typo crashing skyrocket on big endian machines. Spotted by
Michel Danzer.

	Two new savers: biof and busyspheres.

	New texture smooth for colorfire.

	Fixed a bug that left screensavers running after they should have quit.

	Replaced compile time cpu feature detection with runtime detection.

******************************* Version 0.7.0 *********************************

	Extracted all binary data from the source into binary files. The
source distribution has gotten notably smaller as a result. 

	The binary files are now converted during compilation to significantly 
larger source files with escaping instead of base64 type inefficient strings.
The resulting executables have no unnecessary bloat.

	If the bzip2 library is detected on the system, it is used to compress
the binary data before conversion.

	Added --disable-bzip2 flag to configure. This is to give up space in
order to avoid a delay while decompressing, which is noticeable in skyrocket
that has about 3MB of compressed textures and audio.

	Added --disable-sound flag to configure. This chops off about 3.3MB
(before compression) of audio from skyrocket and any runtime overhead
associated with sound. Useful if you intend to keep passing volume = 0 to
skyrocket.

	Added SSE/3DNow optimized code that gives euphoria a nice 10% boost
on my P4, probably a bit more on an Athlon but I haven't had recent console 

src/hufo_tunnel.cpp
/*
 * Copyright (C) 2002  Jeremie Allard (Hufo / N.A.A.)
 * Ported to Linux by Tugrul Galatali <tugrul@galatali.com>
 *
 * hufo_tunnel is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as 
 * published by the Free Software Foundation.
 *
 * hufo_tunnel is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

// hufo_smoke screen saver

#include <math.h>
#include <stdio.h>
#include <GL/gl.h>
#include <GL/glu.h>

#include "driver.h"
#include "hufo_tunnel.h"
#include "hufo_tunnel_textures.h"
#include "loadTexture.h"

const char *hack_name = "Hufo's little tunnel";

int dTexture;
int dWireframe;

#define XSTD (4.0/3.0)
float THole;			// tunnel time
float TVit;

void hack_reshape (xstuff_t * XStuff)
{
	float x = (float)XStuff->windowWidth / (float)XStuff->windowHeight;	// Correct the viewing ratio of the window in the X axis.

	// Window initialization
	glViewport (0, 0, XStuff->windowWidth, XStuff->windowHeight);

	glMatrixMode (GL_PROJECTION);
	glLoadIdentity ();

	if (x > XSTD)
		gluOrtho2D (-x, x, -1, 1);	// Reset to a 2D screen space.
	else
		gluOrtho2D (-XSTD, XSTD, -XSTD / x, XSTD / x);	// Reset to a 2D screen space.

	// Reset model view matrix stack.
	glMatrixMode (GL_MODELVIEW);
	glLoadIdentity ();
}

// Startup Stuff.
void hack_init (xstuff_t * XStuff)	// Called right after the window is created, and OpenGL is initialized.
{
	// Reset the matrix to something we know.
	hack_reshape (XStuff);

	if (!dWireframe) {
		if (dTexture) {
			unsigned char *l_tex;
			unsigned int tex;

			glGenTextures (1, &tex);

			glBindTexture (GL_TEXTURE_2D, tex);

			glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
			glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
			glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
			glTexParameterf (GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_NEAREST);

			if (dTexture == 1) {
				LOAD_TEXTURE (l_tex, swirlmap, swirlmap_compressedsize, swirlmap_size)
				gluBuild2DMipmaps (GL_TEXTURE_2D, 3, 128, 128, GL_RGB, GL_UNSIGNED_BYTE, l_tex);
			} else {
				LOAD_TEXTURE (l_tex, marblemap, marblemap_compressedsize, marblemap_size)
				gluBuild2DMipmaps (GL_TEXTURE_2D, 3, 256, 256, GL_RGB, GL_UNSIGNED_BYTE, l_tex);
			}
			FREE_TEXTURE (l_tex)

				glEnable (GL_TEXTURE_2D);
		}
	}

	glCullFace (GL_FRONT);	// reject fliped faces
	glEnable (GL_CULL_FACE);
	glDisable (GL_DEPTH_TEST);	// no zbuffer

	THole = 0.0;
	TVit = 8000.0;
	HoleInit ();		// initialise tunnel pos
}

void hack_cleanup (xstuff_t * XStuff)
{
}

// Draw all the scene related stuff.
void hack_draw (xstuff_t * XStuff, double currentTime, float frameTime)
{
	THole += frameTime * TVit;
	CalcHole ((int)THole);	// animate the tunnel
	if (HoleNbImgA == 0) {
		glClear (GL_COLOR_BUFFER_BIT);

		return;
	}
	// and render it
	glClear (GL_COLOR_BUFFER_BIT);
	int p, i;
	float f1, f2;

	for (p = HoleNbImgA - 2; p >= 0; --p) {
		f1 = min (1.0, 1.0 / (0.1 + PtDist[p] * (0.15)));
		f2 = min (1.0, 1.0 / (0.1 + PtDist[p + 1] * (0.15)));

		if (dWireframe) {
			glBegin (GL_LINES);
		} else {
			glBegin (GL_QUAD_STRIP);
		}

		for (i = 0; i <= HoleNbParImg; i += ((dCoarse > 0) ? dCoarse : 1)) {
			float f;

			if (dCoarse) {
				f = f1 * Pt[p][i].c1;
				glColor3f (f, f, f);
			} else {
				glColor3f (f1, f1, f1);
			}

			glTexCoord2f (Pt[p][i].u, Pt[p][i].v);
			glVertex2f (Pt[p][i].ex, Pt[p][i].ey);

			if (dCoarse) {
				f = f2 * Pt[p + 1][i].c1;
				glColor3f (f, f, f);
			} else {
				glColor3f (f2, f2, f2);
			}

			glTexCoord2f (Pt[p + 1][i].u, Pt[p + 1][i].v);
			glVertex2f (Pt[p + 1][i].ex, Pt[p + 1][i].ey);
		}
		glEnd ();
		if (dWireframe) {
			if (BBoxEmpty (&BBPlan[p]))
				glColor3f (1.0, 0.0, 0.0);
			else
				glColor3f (0.0, 1.0, 0.0);
			glBegin (GL_LINE_STRIP);
			glVertex2f (BBPlan[p].u0, BBPlan[p].v0);
			glVertex2f (BBPlan[p].u1, BBPlan[p].v0);
			glVertex2f (BBPlan[p].u1, BBPlan[p].v1);
			glVertex2f (BBPlan[p].u0, BBPlan[p].v1);
			glVertex2f (BBPlan[p].u0, BBPlan[p].v0);
			glEnd ();
			glColor3f (f1, f1, f1);
			glBegin (GL_LINE_STRIP);
			for (i = 0; i <= HoleNbParImg; i += ((dCoarse > 0) ? dCoarse : 1)) {
				glTexCoord2f (Pt[p][i].u, Pt[p][i].v);
				glVertex2f (Pt[p][i].ex, Pt[p][i].ey);
			}
			glEnd ();
		}
	}
}

void hack_handle_opts (int argc, char **argv)
{
	int change_flag = 0;

	dTexture = 1;
	dCoarse = 0;
	dSinHole = 0;
	dWireframe = 0;

	while (1) {
		int c;

#ifdef HAVE_GETOPT_H
		static struct option long_options[] = {
			{"help", 0, 0, 'h'},
			DRIVER_OPTIONS_LONG 
			{"texture", 1, 0, 't'},
			{"swirl", 0, 0, 1},
			{"marble", 0, 0, 2},
			{"coarseness", 1, 0, 'c'},
			{"sinusoide", 0, 0, 's'},
			{"no-sinusoide", 0, 0, 'S'},
			{"wireframe", 0, 0, 'w'},
			{"no-wireframe", 0, 0, 'W'},
			{0, 0, 0, 0}
		};

		c = getopt_long (argc, argv, DRIVER_OPTIONS_SHORT "hc:t:sSwW", long_options, NULL);
#else
		c = getopt (argc, argv, DRIVER_OPTIONS_SHORT "hc:t:sSwW");
#endif
		if (c == -1)
			break;

		switch (c) {
		DRIVER_OPTIONS_CASES case 'h':
			printf ("%s:"
#ifndef HAVE_GETOPT_H
				" Not built with GNU getopt.h, long options *NOT* enabled."
#endif
				"\n" DRIVER_OPTIONS_HELP "\t--texture/-t <arg>\n" "\t--swirl\n" "\t--marble\n" "\t--coarseness/-c <arg>\n" "\t--sinusoide/-s\n" "\t--no-sinusoide/-S\n" 
				"\t--wireframe/-w\n" "\t--no-wireframe/-W\n", argv[0]);
			exit (1);
		case 't':
			change_flag = 1;
			dTexture = strtol_minmaxdef (optarg, 10, 0, 2, 0, 1, "--texture: ");
			break;
		case 1:
			change_flag = 1;
			dTexture = 1;
			break;
		case 2:
			change_flag = 1;
			dTexture = 2;
			break;
		case 'c':
			change_flag = 1;
			dCoarse = 1 << (3 - strtol_minmaxdef (optarg, 10, 0, 3, 1, 0, "--coarseness: "));

			if (dCoarse == 8)
				dCoarse = 0;

			break;
		case 's':
			change_flag = 1;
			dSinHole = 1;
			break;
		case 'S':
			change_flag = 1;
			dSinHole = 0;
			break;
		case 'w':
			change_flag = 1;
			dWireframe = 1;
			break;
		case 'W':
			change_flag = 1;
			dWireframe = 0;
			break;
		}
	}

	if (!change_flag) {
		dTexture = rsRandi (2) + 1;
		dCoarse = 1 << (3 - rsRandi (4));

		if (dCoarse == 8)
			dCoarse = 0;

		dSinHole = rsRandi (2);
		dWireframe = (rsRandi (10) == 0);
	}
}
